### PR TITLE
Application forking fixes and improvements

### PR DESCRIPTION
**Pull request by sharat87**
Fixes the following items in the application forking flow:

1. Fix for new users getting datasources of example organization, even though there are no apps using it.
2. When cloning an application to another organization, clone *only* those datasources that are used by that application.
3. Before cloning a datasource to a target organization, check the datasources already present in that organization for a match and, if found, use that instead of cloning.
4. When cloning an application or datasource, if a different application or datasource already exists in the target organization, append an incrementing suffix number like `(1)`, `(2)`, `(3)` etc., instead of throwing up with duplicate key exception.
5. Allow forking of an app even if `forkingEnabled` is `false`, but if the user has manage access on this application. This will enable the use of this API for actions like *copy app to other org* etc.

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java [hunk 1]
```diff
@@ -59,4 +60,24 @@
...
+     */
+    public boolean softEquals(Datasource other) {
+        if (other == null) {
+            return false;
+        }
+
+        return new EqualsBuilder()
+                .append(name, other.name)
+                .append(pluginId, other.pluginId)
+                .append(isAutoGenerated, other.isAutoGenerated)
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java:78]: Should this be applicable only for `isAutogenerated` being false? For embedded datasources, we'd want to generate a new embedded datasource all the time, yes?

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java:78]: Ah thanks for pointing this out. This is mostly off of ignorance on my part. I wasn't really sure what this field does. So just checked for it's equality as well. If it's for embedded datasources, the code won't even check equality. The equality check will only happen with the datasource has an ID, which would be missing in embedded datasources.

That said, if the target organization where we are forking, has a datasource with `isAutoGenerated` as `true`, should we create a new datasource for the forked application? If no, then we can remove this line 78. Am I making sense?

Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java:78]: Not entirely sure if I understood this correctly, you mean if the source application has an embedded datasource then the forked one should be embedded as well? I think yes. Users can save later if required.

Also are doing the check for `id` somewhere upstream?

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java:78]: Correct. We only ever clone datasources, if they have an ID.
Screenshot 2021-03-15 at 18 01 10

Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/domains/Datasource.java:78]: Fair enough, extensible in that sense. A shallow comparison of datasources is only valid between two embedded or two global ones. Closing this.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ExamplesOrganizationCloner.java [hunk 2]
```diff
@@ -326,35 +328,105 @@ private boolean updateOnLoadActionsWithNewActionIds(Map<String, String> actionId
...
-                    if (templateDatasourceId == null) {
-                        return Mono.error(new AppsmithException(AppsmithError.INTERNAL_SERVER_ERROR));
+    private Mono<Datasource> createSuffixedDatasource(Datasource datasource, String name, int suffix) {
+        final String actualName = name + (suffix == 0 ? "" : " (" + suffix + ")");
+        datasource.setName(actualName);
+        return datasourceService.create(datasource)
+                .onErrorResume(DuplicateKeyException.class, error -> {
+                    if (suffix <= CLONE_SUFFIX_LIMIT
+                            && error.getMessage() != null
+                            && error.getMessage().contains("organization_datasource_deleted_compound_index")) {
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ExamplesOrganizationCloner.java:402]: Really minor, but if we're going up to 20 in the handling we should also add handling for what happens when you're 21+ 😄

Comment (nidhi-nair) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ExamplesOrganizationCloner.java:402]: Rest of it LGTM, ping me any time for approval

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/ExamplesOrganizationCloner.java:402]: This I actually later intended to remove it. If you notice, I didn't set any such limit to cloning applications, which uses a very similar flow. I added it when testing a few strategies of doing this, but I don't see much point to it now. I'll remove this and push in a minute.

**Review by nidhi-nair — APPROVED** [review]

**Pull request merged**